File: go/HttpClient.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// HttpClientOption HTTP客户端配置选项
type HttpClientOption struct {
	Header                   map[string]string
	Cookie                   map[string]string
	ProxyAddress             string
	SocketTimeout            int // 毫秒
	ConnectTimeout           int // 毫秒
	IgnoreSSL                bool
	defaultSocketTimeout     int // 秒
	defaultConnectTimeout    int // 秒
	connectionRequestTimeout int // 秒
}

// NewHttpClientOption 创建默认的HTTP客户端配置
func NewHttpClientOption() *HttpClientOption {
	return &HttpClientOption{
		Header:                   make(map[string]string),
		Cookie:                   make(map[string]string),
		IgnoreSSL:                true,
		defaultSocketTimeout:     60,
		defaultConnectTimeout:    60,
		connectionRequestTimeout: 5,
	}
}

// HTTPResponse HTTP响应封装
type HTTPResponse struct {
	RequestMethod string
	RequestURL    string
	StatusCode    int
	Headers       map[string][]string
	Content       []byte
	ElapsedTime   float64 // 毫秒
}

// Text 获取响应文本
func (r *HTTPResponse) Text() string {
	return string(r.Content)
}

// JSON 解析JSON响应
func (r *HTTPResponse) JSON(v interface{}) error {
	return json.Unmarshal(r.Content, v)
}

func (r *HTTPResponse) String() string {
	return fmt.Sprintf("HTTPResponse(status_code=%d, elapsed_time=%.2fms, content_length=%d)",
		r.StatusCode, r.ElapsedTime, len(r.Content))
}

// HttpClient HTTP客户端
type HttpClient struct {
	option *HttpClientOption
	client *http.Client
}

// NewHttpClient 创建HTTP客户端
func NewHttpClient(option *HttpClientOption) *HttpClient {
	if option == nil {
		option = NewHttpClientOption()
	}

	// 创建HTTP客户端
	client := &http.Client{}

	// 配置超时
	connectTimeout := time.Duration(option.ConnectTimeout) * time.Millisecond
	if option.ConnectTimeout == 0 {
		connectTimeout = time.Duration(option.defaultConnectTimeout) * time.Second
	}

	socketTimeout := time.Duration(option.SocketTimeout) * time.Millisecond
	if option.SocketTimeout == 0 {
		socketTimeout = time.Duration(option.defaultSocketTimeout) * time.Second
	}

	client.Timeout = connectTimeout + socketTimeout

	// 配置Transport
	transport := &http.Transport{
		DialContext: (&net.Dialer{
			Timeout:   connectTimeout,
			KeepAlive: 30 * time.Second,
		}).DialContext,
		MaxIdleConns:          200,
		MaxIdleConnsPerHost:   20,
		IdleConnTimeout:       90 * time.Second,
		TLSHandshakeTimeout:   10 * time.Second,
		ExpectContinueTimeout: 1 * time.Second,
		MaxConnsPerHost:       20,
	}

	// 配置代理
	if option.ProxyAddress != "" {
		proxyURL, err := url.Parse("http://" + option.ProxyAddress)
		if err != nil {
			panic(fmt.Sprintf("无效的代理地址: %v", err))
		}
		transport.Proxy = http.ProxyURL(proxyURL)
	}

	// 配置SSL验证
	if option.IgnoreSSL {
		transport.TLSClientConfig = &tls.Config{
			InsecureSkipVerify: true,
		}
	}

	client.Transport = transport

	return &HttpClient{
		option: option,
		client: client,
	}
}

// buildRequest 构建基础请求
func (c *HttpClient) buildRequest(method, urlStr string, body io.Reader, headers map[string]string) (*http.Request, error) {
	req, err := http.NewRequest(method, urlStr, body)
	if err != nil {
		return nil, err
	}

	// 设置默认头
	for k, v := range c.option.Header {
		req.Header.Set(k, v)
	}

	// 设置额外头
	if headers != nil {
		for k, v := range headers {
			req.Header.Set(k, v)
		}
	}

	// 设置Cookie
	for k, v := range c.option.Cookie {
		req.AddCookie(&http.Cookie{Name: k, Value: v})
	}

	return req, nil
}

// doRequest 执行请求并返回响应
func (c *HttpClient) doRequest(req *http.Request) (*HTTPResponse, error) {
	startTime := time.Now()

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// 读取响应内容
	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// 计算耗时(毫秒)
	elapsedTime := float64(time.Since(startTime).Microseconds()) / 1000

	return &HTTPResponse{
		RequestMethod: req.Method,
		RequestURL:    req.URL.String(),
		StatusCode:    resp.StatusCode,
		Headers:       resp.Header,
		Content:       content,
		ElapsedTime:   elapsedTime,
	}, nil
}

// Get 发送GET请求
func (c *HttpClient) Get(urlStr string, headers map[string]string) (*HTTPResponse, error) {
	req, err := c.buildRequest("GET", urlStr, nil, headers)
	if err != nil {
		return nil, err
	}

	return c.doRequest(req)
}

// PostJSON 发送POST JSON请求
func (c *HttpClient) PostJSON(urlStr string, data interface{}, headers map[string]string) (*HTTPResponse, error) {
	// 序列化JSON数据
	var jsonData []byte
	var err error

	switch v := data.(type) {
	case string:
		jsonData = []byte(v)
	default:
		jsonData, err = json.Marshal(v)
		if err != nil {
			return nil, err
		}
	}

	// 准备请求头
	reqHeaders := make(map[string]string)
	reqHeaders["Content-Type"] = "application/json; charset=utf-8"

	// 合并传入的头
	if headers != nil {
		for k, v := range headers {
			reqHeaders[k] = v
		}
	}

	req, err := c.buildRequest("POST", urlStr, bytes.NewBuffer(jsonData), reqHeaders)
	if err != nil {
		return nil, err
	}

	return c.doRequest(req)
}

// PostForm 发送POST表单请求
func (c *HttpClient) PostForm(urlStr string, data map[string]string, headers map[string]string) (*HTTPResponse, error) {
	// 准备表单数据
	formData := url.Values{}
	for k, v := range data {
		formData.Set(k, v)
	}

	// 准备请求头
	reqHeaders := make(map[string]string)
	reqHeaders["Content-Type"] = "application/x-www-form-urlencoded; charset=UTF-8"

	// 合并传入的头
	if headers != nil {
		for k, v := range headers {
			reqHeaders[k] = v
		}
	}

	req, err := c.buildRequest("POST", urlStr, strings.NewReader(formData.Encode()), reqHeaders)
	if err != nil {
		return nil, err
	}

	return c.doRequest(req)
}

// GetJSON 发送带JSON体的GET请求
func (c *HttpClient) GetJSON(urlStr string, data interface{}) (*HTTPResponse, error) {
	// 序列化JSON数据
	var jsonData []byte
	var err error

	switch v := data.(type) {
	case string:
		jsonData = []byte(v)
	default:
		jsonData, err = json.Marshal(v)
		if err != nil {
			return nil, err
		}
	}

	headers := map[string]string{
		"Content-Type": "application/json; charset=utf-8",
	}

	req, err := c.buildRequest("GET", urlStr, bytes.NewBuffer(jsonData), headers)
	if err != nil {
		return nil, err
	}

	return c.doRequest(req)
}

// File 用于上传的文件结构
type File struct {
	Filename string
	Content  []byte
}

// UploadFiles 上传文件
func (c *HttpClient) UploadFiles(urlStr string, files []File) (*HTTPResponse, error) {
	// 构建multipart/form-data
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	for i, file := range files {
		part, err := writer.CreateFormFile(fmt.Sprintf("file_%d", i), file.Filename)
		if err != nil {
			return nil, err
		}
		_, err = part.Write(file.Content)
		if err != nil {
			return nil, err
		}
	}

	err := writer.Close()
	if err != nil {
		return nil, err
	}

	headers := map[string]string{
		"Content-Type": writer.FormDataContentType(),
	}

	req, err := c.buildRequest("POST", urlStr, body, headers)
	if err != nil {
		return nil, err
	}

	return c.doRequest(req)
}

// Close 关闭客户端（在Go中HTTP客户端通常不需要显式关闭）
func (c *HttpClient) Close() {
	// 对于http.Client，没有需要显式关闭的资源
	// 如果需要关闭连接，可以配置Transport的CloseIdleConnections
	if transport, ok := c.client.Transport.(*http.Transport); ok {
		transport.CloseIdleConnections()
	}
}

File: python/src/HttpClient.py
# HttpClient.py
import json
import ssl
import time
from typing import Dict, List, Optional, Any, Union
from urllib.parse import urlparse
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.ssl_ import create_urllib3_context
from requests.packages.urllib3.poolmanager import PoolManager


class HttpClientOption:
    """HTTP 客户端配置选项"""

    def __init__(self,
                 header: Optional[Dict[str, Any]] = None,
                 cookie: Optional[Dict[str, Any]] = None,
                 proxy_address: Optional[str] = None,
                 socket_timeout: Optional[int] = None,
                 connect_timeout: Optional[int] = None,
                 ignore_ssl: bool = True):
        """
        初始化 HttpClient 配置

        Args:
            header: 请求头字典
            cookie: Cookie 字典
            proxy_address: 代理地址 (格式: "host:port")
            socket_timeout: socket 超时时间（毫秒）
            connect_timeout: 连接超时时间（毫秒）
            ignore_ssl: 是否忽略 SSL 证书验证
        """
        self.header = header or {}
        self.cookie = cookie or {}
        self.proxy_address = proxy_address
        self.socket_timeout = socket_timeout
        self.connect_timeout = connect_timeout
        self.ignore_ssl = ignore_ssl

        # 默认超时设置（单位：秒）
        self.default_socket_timeout = 60  # 60秒
        self.default_connect_timeout = 60  # 60秒
        self.connection_request_timeout = 5  # 5秒


class IgnoreSSLAdapter(HTTPAdapter):
    """忽略 SSL 证书验证的适配器"""

    def init_poolmanager(self, *args, **kwargs):
        ctx = create_urllib3_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
        kwargs['ssl_context'] = ctx
        return super().init_poolmanager(*args, **kwargs)


class HTTPResponse:
    """HTTP 响应封装类"""

    def __init__(self, request_method: str, request_url: str,
                 status_code: int, headers: Dict[str, str],
                 content: bytes, elapsed_time: float):
        """
        初始化 HTTP 响应

        Args:
            request_method: 请求方法
            request_url: 请求URL
            status_code: 状态码
            headers: 响应头
            content: 响应内容
            elapsed_time: 请求耗时（毫秒）
        """
        self.request_method = request_method
        self.request_url = request_url
        self.status_code = status_code
        self.headers = headers
        self.content = content
        self.elapsed_time = elapsed_time

    @property
    def text(self) -> str:
        """获取响应文本"""
        return self.content.decode('utf-8', errors='ignore')

    def json(self) -> Any:
        """解析 JSON 响应"""
        return json.loads(self.text)

    def __str__(self) -> str:
        return (f"HTTPResponse(status_code={self.status_code}, "
                f"elapsed_time={self.elapsed_time}ms, "
                f"content_length={len(self.content)})")


class HttpClient:
    """HTTP 客户端类"""

    def __init__(self, option: Optional[Union[Dict[str, Any], HttpClientOption]] = None):
        """
        初始化 HTTP 客户端

        Args:
            option: 配置选项，可以是字典或 HttpClientOption 对象
        """
        if option is None:
            option = {}

        # 转换配置选项
        if isinstance(option, dict):
            self.option = HttpClientOption(
                header=option.get('header'),
                cookie=option.get('cookie'),
                proxy_address=option.get('proxy_address'),
                socket_timeout=option.get('socket_timeout'),
                connect_timeout=option.get('connect_timeout'),
                ignore_ssl=option.get('ignore_ssl', True)
            )
        else:
            self.option = option

        # 创建会话
        self.session = requests.Session()

        # 设置默认请求头
        if self.option.header:
            self.session.headers.update(
                {k: str(v) for k, v in self.option.header.items()}
            )

        # 设置 Cookie
        if self.option.cookie:
            for key, value in self.option.cookie.items():
                self.session.cookies.set(key, str(value))

        # 设置代理
        if self.option.proxy_address:
            proxy_parts = self.option.proxy_address.split(':')
            if len(proxy_parts) != 2:
                raise ValueError("请输入正确的proxyAddress地址 (格式: host:port)")

            proxy_host = proxy_parts[0]
            proxy_port = int(proxy_parts[1])
            self.session.proxies = {
                'http': f'http://{proxy_host}:{proxy_port}',
                'https': f'http://{proxy_host}:{proxy_port}'
            }

        # 设置 SSL 验证
        self.session.verify = not self.option.ignore_ssl

        # 添加忽略 SSL 的适配器
        if self.option.ignore_ssl:
            adapter = IgnoreSSLAdapter()
            self.session.mount('https://', adapter)
            self.session.mount('http://', adapter)

        # 设置连接池
        adapter = HTTPAdapter(
            pool_connections=200,
            pool_maxsize=20,
            max_retries=3
        )
        self.session.mount('https://', adapter)
        self.session.mount('http://', adapter)

    def _build_timeout(self) -> tuple:
        """构建超时配置"""
        connect_timeout = (
            self.option.connect_timeout / 1000
            if self.option.connect_timeout
            else self.option.default_connect_timeout
        )

        read_timeout = (
            self.option.socket_timeout / 1000
            if self.option.socket_timeout
            else self.option.default_socket_timeout
        )

        return (connect_timeout, read_timeout)

    def _build_response(self, response: requests.Response,
                        start_time: float) -> HTTPResponse:
        """构建 HTTPResponse 对象"""
        elapsed_time = (time.time() - start_time) * 1000  # 转换为毫秒

        return HTTPResponse(
            request_method=response.request.method,
            request_url=response.request.url,
            status_code=response.status_code,
            headers=dict(response.headers),
            content=response.content,
            elapsed_time=elapsed_time
        )

    def get(self, url: str, headers: Optional[Dict[str, Any]] = None) -> HTTPResponse:
        """
        GET 请求

        Args:
            url: 请求URL
            headers: 请求头

        Returns:
            HTTPResponse 对象
        """
        start_time = time.time()
        timeout = self._build_timeout()

        response = self.session.get(
            url,
            headers=headers,
            timeout=timeout
        )

        return self._build_response(response, start_time)

    def post_json(self, url: str, data: Any,
                  headers: Optional[Dict[str, Any]] = None) -> HTTPResponse:
        """
        POST JSON 请求

        Args:
            url: 请求URL
            data: 请求数据（可以是字典或 JSON 字符串）
            headers: 请求头

        Returns:
            HTTPResponse 对象
        """
        start_time = time.time()
        timeout = self._build_timeout()

        # 准备请求头
        request_headers = {'Content-Type': 'application/json; charset=utf-8'}
        if headers:
            request_headers.update(headers)

        # 准备请求数据
        if isinstance(data, str):
            json_data = data
        else:
            json_data = json.dumps(data, ensure_ascii=False)

        response = self.session.post(
            url,
            data=json_data,
            headers=request_headers,
            timeout=timeout
        )

        return self._build_response(response, start_time)

    def post_form(self, url: str, data: Dict[str, Any],
                  headers: Optional[Dict[str, Any]] = None) -> HTTPResponse:
        """
        POST 表单请求

        Args:
            url: 请求URL
            data: 表单数据
            headers: 请求头

        Returns:
            HTTPResponse 对象
        """
        start_time = time.time()
        timeout = self._build_timeout()

        # 准备请求头
        request_headers = {
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
        }
        if headers:
            request_headers.update(headers)

        response = self.session.post(
            url,
            data=data,
            headers=request_headers,
            timeout=timeout
        )

        return self._build_response(response, start_time)

    def get_json(self, url: str, data: Any) -> HTTPResponse:
        """
        GET 请求（带 JSON 请求体）

        Args:
            url: 请求URL
            data: 请求数据（可以是字典或 JSON 字符串）

        Returns:
            HTTPResponse 对象
        """
        start_time = time.time()
        timeout = self._build_timeout()

        # 准备请求数据
        if isinstance(data, str):
            json_data = data
        else:
            json_data = json.dumps(data, ensure_ascii=False)

        # 使用 requests 的 request 方法发送带请求体的 GET 请求
        response = self.session.request(
            'GET',
            url,
            data=json_data,
            headers={'Content-Type': 'application/json; charset=utf-8'},
            timeout=timeout
        )

        return self._build_response(response, start_time)

    def upload_files(self, url: str, files: List[Any]) -> HTTPResponse:
        """
        上传文件

        Args:
            url: 上传URL
            files: 文件列表

        Returns:
            HTTPResponse 对象
        """
        start_time = time.time()
        timeout = self._build_timeout()

        # 准备文件字典
        files_dict = {}
        for i, file in enumerate(files):
            # 假设 file 有 filename 和 content 属性
            # 根据实际情况调整
            if hasattr(file, 'filename') and hasattr(file, 'content'):
                files_dict[f'file_{i}'] = (file.filename, file.content)
            else:
                # 如果文件是字典形式
                files_dict[f'file_{i}'] = (file.get('filename'), file.get('content'))

        response = self.session.post(
            url,
            files=files_dict,
            timeout=timeout
        )

        return self._build_response(response, start_time)

    def close(self) -> None:
        """关闭 HTTP 客户端"""
        if self.session:
            self.session.close()

    def __enter__(self):
        """上下文管理器入口"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """上下文管理器退出"""
        self.close()


